bot/bot.go

```diff
@@ -64,26 +46,33 @@ func New(cfg *config.Config, dal *data.DAL, gh *github.API, log *log.Entry) *Bot
 	api := slack.New(cfg.SlackToken)
 
 	b := &Bot{
-		token: cfg.SlackToken,
-		api:   api,
-		rtm:   api.NewRTM(),
-		dal:   dal,
-		gh:    gh,
-		log:   log,
+		token:    cfg.SlackToken,
+		api:      api,
+		rtm:      api.NewRTM(),
+		dal:      dal,
+		gh:       gh,
+		log:      log,
+		commands: map[string]*cmd.Command{},
 	}
+	b.PopulateUsers()
 
-	commands := map[string]CommandHandler{
-		"help":    b.help,
-		"me":      b.me,
-		"set":     b.set,
-		"add":     b.add,
-		"remove":  b.remove,
-		"view":    b.view,
-		"refresh": b.refresh,
+	// Attach command handlers
+	for _, cmd := range Commands {
+		b.commands[cmd.Name] = cmd
 	}
-	b.commands = commands
 
-	b.PopulateUsers()
+	HelpCmd.HandleFunc = b.help
+	SetCmd.HandleFunc = b.set
+	ViewUserCmd.HandleFunc = b.viewUser
+	ViewTeamCmd.HandleFunc = b.viewTeam
+	AddUserCmd.HandleFunc = b.addUser
+	AddTeamCmd.HandleFunc = b.addTeam
+	AddAdminCmd.HandleFunc = b.addAdmin
+	RemoveAdminCmd.HandleFunc = b.removeAdmin
+	RemoveUserCmd.HandleFunc = b.removeUser
+	RemoveTeamCmd.HandleFunc = b.removeTeam
+	TeamsCmd.HandleFunc = b.listTeams
+	RefreshCmd.HandleFunc = b.refresh
```

Comment:
Seems a bit tedious to go through this flow for implementing new commands:
1. add a command definition in `commands`
2. add definition to list of commands in `commands.Commands`
3. write a handler func in `bot`
4. add to a growing list of `SomeCmd.HandleFunc = b.myHandler` assignments in `bot`

Would it be better to have the `HandleFunc` for each command declared in the `command` package and have the bot pass itself as a parameter?

```go
func (b *Bot) handleMessageEvent(msg slack.Msg) {
    // ...
    res, params, err := cmd.Execute(context, b)
    // ...
}
```

Reply:
You're right, this has actually been troubling me as well. Passing the bot as a parameter is okay, but it might be better to declare the commands as part of the bot's definition, or in it's `New` function. I'll think about it and get back to you.